File: other-programming-languages/Ruby/2024/2024-002/safe-reports.rb
# Function to check if a report is safe
def is_safe(report)
  differences = report.each_cons(2).map { |a, b| b - a }
  all_increasing = differences.all? { |diff| diff >= 1 && diff <= 3 }
  all_decreasing = differences.all? { |diff| diff <= -1 && diff >= -3 }
  all_increasing || all_decreasing
end

# Function to check if a report is safe with the Problem Dampener
def is_safe_with_dampener(report)
  return true if is_safe(report)

  report.each_index do |i|
    modified_report = report[0...i] + report[i+1..-1]
    return true if is_safe(modified_report)
  end

  false
end

# Main function to process the input and count safe reports
def main
  input_path = "\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-002\\input.txt"
  
  # Read and process the input file
  reports = File.readlines(input_path).map do |line|
    line.split.map(&:to_i)
  end

  # Count the safe reports
  safe_count = reports.count { |report| is_safe(report) }
  puts "Safe reports: #{safe_count}"

  # Count the safe reports with the Problem Dampener
  safe_with_dampener_count = reports.count { |report| is_safe_with_dampener(report) }
  puts "Safe reports with dampener: #{safe_with_dampener_count}"
end

# Run the main function
main

File: other-programming-languages/Typescript/2024/2024-002/safeReports.ts
import * as fs from "fs";
import * as path from "path";

// Function to check if a report is safe
function isSafe(report: number[]): boolean {
    const differences = report.slice(1).map((num, i) => num - report[i]);

    const allIncreasing = differences.every(diff => diff >= 1 && diff <= 3);
    const allDecreasing = differences.every(diff => diff <= -1 && diff >= -3);

    return allIncreasing || allDecreasing;
}

// Function to check if a report is safe with the Problem Dampener
function isSafeWithDampener(report: number[]): boolean {
    if (isSafe(report)) {
        return true;
    }

    for (let i = 0; i < report.length; i++) {
        const modifiedReport = report.slice(0, i).concat(report.slice(i + 1));
        if (isSafe(modifiedReport)) {
            return true;
        }
    }

    return false;
}

// Main function to process the input and count safe reports
function main() {
    const inputPath = path.resolve("\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-002\\input.txt");

    // Read the input file
    const data = fs.readFileSync(inputPath, "utf-8");
    const lines = data.trim().split("\n");

    // Convert input data into an array of reports
    const reports = lines.map(line => line.split(/\s+/).map(Number));

    // Count the number of safe reports
    const safeCount = reports.filter(isSafe).length;
    console.log(`Safe reports: ${safeCount}`);

    // Count the number of safe reports with the Problem Dampener
    const safeWithDampenerCount = reports.filter(isSafeWithDampener).length;
    console.log(`Safe reports with dampener: ${safeWithDampenerCount}`);
}

// Run the main function
main();
